### DISCUSSION TITLE
ipxe and dhcp ipv6

### DISCUSSION THREAD

**Discussion by tokenrain** [ipxe/ipxe#839] [Q&A]
I have a set of ubuntu linux boxes acting as simple kvm hypervisors via qemu and bridged networking. I build most of my boxes via pxe and kickstart/preseed. Whenever I boot a vm domain off the network to build the dhcp process takes ~17 seconds. 1 second for ipv4 dhcp and then the rest is ipxe doing attempting dhcp for ipv6 and finally timing out.

I proved this out by booting to the ipxe command line and running:
dhcp (17 seconds)
dhcp -c dhcp net0 (<1 sec)
dhcp -c ipv6 net0 (17 seconds)

As expected after this I have an ipv4 address but no ipv6. The hypervisor has ipv6 disabled but the router on the network supports ipv6 multicast

Other than rebuilding ipxe without ipv6 support does anyone know of a way to prevent ipxe from performing dhcpv6? Thanks!

```
ipxe-qemu                       1.0.0+git-20190109.133f4c4-0ubuntu3.2
ipxe-qemu-256k-compat-efi-roms  1.0.0+git-20150424.a25a16d-0ubuntu4

------------------------------------

iPXE 1.0.0+git-20190109.133f4c4-0ubuntu3.2 -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP HTTPS iSCSI NFS TFTP AoE ELF MBOOT PXE bzImage Menu PXEXT

net0: 52:54:00:8a:4d:a1 using virtio-net on 0000:00:02.0 (open)
  [Link:up, TX:0 TXE:0 RX:0 RXE:0]
Configuring (net0 52:54:00:8a:4d:a1)...................

------------------------------------

iPXE> route                              
net0: fe80::5054:ff:fe8a:4da1/64 (inaccessible)

iPXE> dhcp -c dhcp net0
Configuring [dhcp] (net0 52:54:00:8a:4d:a1)... ok

iPXE> route             

net0: 10.0.0.131/255.255.255.0 gw 10.0.0.1
net0: fe80::5054:ff:fe8a:4da1/64

iPXE> dhcp -c ipv6 net0  
Configuring [ipv6] (net0 52:54:00:8a:4d:a1).................. Connection timed out (http://ipxe.org/4c076035)
```

**Comment by NiKiZe** [comment]
Do you have any v6 RAs configured?
iPXE will try to get the network fully working before going further. If you have RA it can be made to not wait for DHCP(v6) another option would be to set "no v6" flag in dhcpv4. (Theoretically, might not be implemented in iPXE)

Another approach would be to have iPXE do this in the background and defer the actual wait to when needed. But that would be a breaking change, and add complexity, not just for iPXE itself, but also for any script that needs to act on such failure.

To answer your question how to prevent V6 without disable v6. You could have embedded script, or use autoexec iPXE feature and there only do V4 DHCP. (I do recommend having RA tho)

**Reply by tokenrain** [reply]
I do not think there are any RAs on this subnet but that is something that I can work with. I will let you know if this is the solution. Thanks for taking the time out to respond.

**Comment by stappersg** [comment]
On Tue, Dec 13, 2022 at 05:17:41PM -0800, Christian I. Nilsson wrote:
> Do you have any v6 RAs configured?

And  "v6 RA" being  IPv6 Router Advertisement.

Examples of software providing  R.A.   are `dnsmasq` and `radvd`.


Groeten
Geert Stappers
-- 
Silence is hard to parse
